tests/e2e/ctl_v3_member_test.go

```diff
@@ -55,6 +58,85 @@ func TestCtlV3MemberUpdatePeerTLS(t *testing.T) {
 	testCtl(t, memberUpdateTest, withCfg(*e2e.NewConfigPeerTLS()))
 }
 
+func TestCtlV3ConsistentMemberList(t *testing.T) {
+	e2e.BeforeTest(t)
+
+	epc, err := e2e.NewEtcdProcessCluster(context.TODO(), t,
+		e2e.WithClusterSize(1),
+		e2e.WithWaitClusterReadyTimeout(0),
+		e2e.WithEnvVars(map[string]string{"GOFAIL_FAILPOINTS": `beforeApplyOneConfChange=sleep("2s")`}),
+	)
+	require.NoError(t, err, "failed to start etcd cluster: %v", err)
+	defer func() {
+		derr := epc.Close()
+		require.NoError(t, derr, "failed to close etcd cluster: %v", derr)
+	}()
+
+	t.Log("Adding and then removing a learner")
+	resp, err := epc.Etcdctl().MemberAddAsLearner(context.TODO(), "newLearner", []string{fmt.Sprintf("http://localhost:%d", e2e.EtcdProcessBasePort+11)})
+	require.NoError(t, err)
+	_, err = epc.Etcdctl().MemberRemove(context.TODO(), resp.Member.ID)
+	require.NoError(t, err)
+	t.Logf("Added and then removed a learner with ID: %x", resp.Member.ID)
+
+	t.Log("Restarting the etcd process to ensure all data is persisted")
+	err = epc.Procs[0].Restart(context.TODO())
+	require.NoError(t, err)
+
+	var wg sync.WaitGroup
+	wg.Add(2)
+	stopc := make(chan struct{}, 2)
+
+	t.Log("Starting a goroutine to repeatedly restart etcdserver")
+	go func() {
+		defer func() {
+			stopc <- struct{}{}
+			wg.Done()
+		}()
+		for i := 0; i < 3; i++ {
+			select {
+			case <-stopc:
+				return
+			default:
+			}
+
+			merr := epc.Procs[0].Restart(context.TODO())
+			require.NoError(t, merr)
+			epc.WaitLeader(t)
+
+			time.Sleep(100 * time.Millisecond)
+		}
+	}()
+
+	t.Log("Starting a goroutine to repeated check the member list")
+	count := 0
+	go func() {
+		defer func() {
+			stopc <- struct{}{}
+			wg.Done()
+		}()
+
+		for {
+			select {
+			case <-stopc:
+				return
+			default:
+			}
+
+			mresp, merr := epc.Etcdctl().MemberList(context.TODO(), true)
```

Comment:
Recommend using `clientv3` library as it's much faster. It should allow much higher chance of reproducing the issue.

Reply:
I am afraid I do not agree this.  An e2e test is supposed to use etcdctl to communicate with etcdserver. Pls follow https://github.com/etcd-io/etcd/pull/16658#discussion_r1341226598 to reproduce this issue locally.

Reply:
No, when testing etcd standard features like Get request, e2e tests should test `etcdctl get`, however using `etcdctl` is not a rule for all tests. There is no benefit of using `etcdctl` for e2e tests, specifically for correctness testing.